streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java

```diff
@@ -333,11 +333,11 @@ public Subscription subscription(Set<String> topics) {
                                     } else {
                                         numPartitionsCandidate = metadata.partitionCountForTopic(sourceTopicName);
                                         if (numPartitionsCandidate == null) {
-                                            repartitionTopicMetadata.get(topicName).numPartitions = NOT_AVAILABLE;
+                                            numPartitionsCandidate = NOT_AVAILABLE;
                                         }
                                     }
 
-                                    if (numPartitionsCandidate != null && numPartitionsCandidate > numPartitions) {
+                                    if (numPartitionsCandidate == NOT_AVAILABLE || numPartitionsCandidate > numPartitions) {
```

Comment:
I don't quite get this yet. If we assign NOT_AVAILABLE to numPartitions as a result of this, the check in the code below will fail:
`if (numPartitions == UNKNOWN)` 

What's the expected behavior if the loop is broken with NOT_AVAILABLE?


Any chance you can add a comment here explaining what's going on?